compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingInterface.cpp

```diff
@@ -390,45 +478,193 @@ FailureOr<TilingResult> TileDispatchUsingSCFForOp::returningMatchAndRewrite(
+  SmallVector<Operation *> tiledAndFusedOps;
+  SmallVector<scf::ForOp> loops;
+};
+
+struct TileAndFuseDispatchUsingSCFForOp
+    : public OpInterfaceRewritePattern<TilingInterface> {
+  /// Construct a generic pattern applied to all TilingInterface ops.
+  TileAndFuseDispatchUsingSCFForOp(MLIRContext *context,
+                                   linalg::LinalgTilingOptions options,
+                                   linalg::LinalgTransformationFilter filter,
+                                   PatternBenefit benefit = 1)
+      : OpInterfaceRewritePattern<TilingInterface>(context, benefit),
+        tilingPattern(context, options, filter) {}
+
+  /// Construct a generic pattern applied to `opName`.
+  TileAndFuseDispatchUsingSCFForOp(StringRef opName, MLIRContext *context,
+                                   linalg::LinalgTilingOptions options,
+                                   linalg::LinalgTransformationFilter filter,
+                                   PatternBenefit benefit = 1)
+      : OpInterfaceRewritePattern<TilingInterface>(context, benefit),
+        tilingPattern(context, options, filter) {}
+
+  FailureOr<TileAndFuseResult> returningMatchAndRewrite(
+      TilingInterface op, PatternRewriter &rewriter) const;
+
+  LogicalResult matchAndRewrite(TilingInterface op,
+                                PatternRewriter &rewriter) const override {
+    return returningMatchAndRewrite(op, rewriter);
+  }
+
+ private:
+  TileDispatchUsingSCFForOp tilingPattern;
+};
+}  // namespace
+
+/// Find all producers to fuse and return then in sorted order;
+static std::vector<Operation *> getAllFusableProducers(TilingInterface op) {
+  llvm::SetVector<Operation *> producers;
+  std::deque<Operation *> worklist;
+  worklist.push_back(op);
+
+  while (!worklist.empty()) {
+    Operation *currOp = worklist.front();
+    worklist.pop_front();
+    for (OpOperand &operand : currOp->getOpOperands()) {
+      auto tilingInterfaceProducer =
+          operand.get().getDefiningOp<TilingInterface>();
+      if (!tilingInterfaceProducer ||
+          producers.count(tilingInterfaceProducer)) {
+        continue;
+      }
+      worklist.push_back(tilingInterfaceProducer);
+      producers.insert(tilingInterfaceProducer);
     }
   }
-  for (auto storeOp : storeOps) {
-    OpResult result = storeOp.getValue().cast<OpResult>();
-    SmallVector<OpFoldResult> resultOffsets, resultSizes;
-    if (failed(op.getResultTilePosition(rewriter, result.getResultNumber(),
-                                        offsets, sizes, resultOffsets,
-                                        resultSizes))) {
-      return rewriter.notifyMatchFailure(
-          op, "failed to rewrite destructive update");
+
+  std::vector<Operation *> sortedOps = sortOpsTopologically(producers);
+  return sortedOps;
+}
+
+/// Return all slices that are used to access a tile of the producer. Assume
+/// that `tiledOps` are in "reverse" order of their appearance in the IR.
```

Comment:
Maybe add a comment to indicate that the last slice op in the returned list will dominate all the other slice ops?